ccp/lib/clients/openshift/client.py

```diff
@@ -0,0 +1,246 @@
+        """
+        Gets service account token assuming user is already logged in
+        :param namespace: The namespace of the service account.
+        :type namespace str
+        :param sa: Default sa/jenkins: The service account whose token is
+        needed
+        :type sa str
+        :return: The token, it was able to successfully retrieve it
+        :rtype: str
+        :raises Exception
+        """
+        c1 = "%s get %s -n %s --template='{{range .secrets}}" % \
+             (self.base_command, sa, namespace)
+        c1 = c1 + "{{ .name }} {{end}}' "
+        c2 = "xargs -n 1 %s get secret -n %s --template=" \
+             % (self.base_command, namespace)
+        c2 = c2 + "'{{ if .data.token }}{{ .data.token }}{{end}}'"
+        c3 = "head -n 1"
+        c4 = "base64 -d -"
+        command = "{} | {} | {} | {}".format(
+            c1, c2, c3, c4
+        )
+        return self.run_command(command, shell=True)
+
+    @retry(tries=5, delay=3, backoff=3)
+    def get_token_from_mounted_secret(
+            self, secret_mount_path="/tmp/jenkins-secret"
+    ):
+        token_location = secret_mount_path + "/token"
+        if not path.exists(token_location):
+            raise Exception(
+                "Could not find token at {}".format(
+                    secret_mount_path
+                )
+            )
+        with open(token_location, "r") as f:
+            return "".join(f.readlines())
+
+    def login(self, server=None, token=None, username=None, password=None):
+        """
+        Logs into an openshift cluster with token or username and password.
+        :param server: Default None: The openshift server to log into. If none
+        is provided then we assume oc is already aware which server it needs
+        to log into
+        :type server str
+        :param token: Default None: The token to login with. Overrides username
+        and password as preferred login method
+        :type token str
+        :param username: Default none : The username to login with. Use only if
+        token is not used.
+        :type username str
+        :param password: The password to login with. Use only if token is not
+        provided. Not recommended
+        :type password str
+        :return: The output ot the command
+        :rtype str
+        :raises Exception
+        """
+        command = str.format(
+            "{base_command} login{token_param}{user_param}{server_param}",
+            base_command=self.base_command,
+            token_param="" if not token else " --token={}".format(
+                token
+            ),
+            user_param="" if token else " -u {} -p {}".format(
+                username,
+                password
+            ),
+            server_param="" if not server else str(server)
+        )
+        return self.run_command(cmd=command)
+
+    @retry(tries=10, delay=3, backoff=2)
+    def process_template(
```

Comment:
Out of curiosity, where do we need this function? Can you cite an example? Because processing a template would mean we make changes to the underlying cluster. However, AFAIK, we're trying to fetch the info from Jenkins and OpenShift as a part of this effort.

Reply:
In the index reader, where we have explicit strings scattered across it for applying seed job and weekly scan job etc. Atleast that is what i remember off the top of my head.

Reply:
This PR does include stuff that is not immediate requirement. But can be used as we go forward, post the API piece. I added is as i did not have anything else to do anyway. And as already said, we should do anyway.

Also, this does not conflict with the title of PR anycase.